[java] AssertionError when outer class is parsed before inner class with conflicting visibility

## Affected version

- PMD 7.24.0
- Current `main` / 7.27.0-SNAPSHOT at `8c0a935`

## Description

When PMD loads an Android dependency processed by R8, `ClassStub` can merge contradictory visibility flags for the same nested class into `public private`. A later symbol-table accessibility check treats this combination as unreachable and throws an `AssertionError`, aborting analysis of the source file.

This is related to #5283 and #5310, but exercises the opposite lazy-loading order. The fix in #5310 removes `public` when a non-public `InnerClasses` entry is merged after the class header. It does not handle the outer class being parsed first.

In the failing order, calls for one `ClassStub` are equivalent to:

```java
inner.setModifiers(Opcodes.ACC_PRIVATE, false);                  // outer InnerClasses entry
inner.setModifiers(Opcodes.ACC_PUBLIC | Opcodes.ACC_SUPER, true); // inner ClassInfo
inner.setModifiers(Opcodes.ACC_PUBLIC, false);                   // inner self-entry
```

The result is `3` (`public private`) instead of `2` (`private`).

## Minimal bytecode shape

The issue can be reproduced without the original dependency by generating two classes with ASM:

- `Outer.class`: public class; its `InnerClasses` entry declares `Outer$Inner` private.
- `Outer$Inner.class`: public class header; its own `InnerClasses` entry declares itself public.
- Resolve/parse `Outer` before `Outer$Inner`.

This models the inconsistent but parseable metadata observed in an R8-processed Android AAR. No source-level Java declaration can express the contradictory metadata.

## Exception

```text
java.lang.AssertionError: This should be unreachable: private field of an interface? ... modifiers: public private static
    at net.sourceforge.pmd.util.AssertionUtil.shouldNotReachHere(AssertionUtil.java:243)
    at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.isAccessible(JavaResolvers.java:479)
    at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.processDeclarations(JavaResolvers.java:432)
    at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.walkSelf(JavaResolvers.java:409)
```

## Expected behavior

Merging class-header and `InnerClasses` access flags should be independent of lazy class parsing order and should never retain more than one visibility flag. Consistent with #5310, the non-public `InnerClasses` visibility should take precedence over `ACC_PUBLIC`.

`pmd.error_recovery=true` limits the impact by reporting a processing error and skipping the affected source file, but it is not a complete fix because all PMD rules are then skipped for that file.

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -79,6 +79,7 @@ Note: Type data is not yet accessible in XPath rules or the PMD Rule Designer. T
 * java
     * [#5041](https://github.com/pmd/pmd/issues/5041): \[java] Parsing failed in ParseLock#doParse(): IndexOutOfBoundsException 
     * [#6768](https://github.com/pmd/pmd/issues/6768): \[java] Disambiguation IllegalStateException resolving a synthesized record accessor used as a call argument alongside an anonymous class
+    * [#6932](https://github.com/pmd/pmd/issues/6932): \[java] AssertionError when outer class is parsed before inner class with conflicting visibility
 * java-bestpractices
     * [#2033](https://github.com/pmd/pmd/issues/2033): \[jsp] NoClassAttribute rule is generating a false positive
     * [#5514](https://github.com/pmd/pmd/issues/5514): \[java] ExhaustiveSwitchHasDefault fails for non-exhaustive switch statements

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ClassStub.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ClassStub.java
@@ -48,6 +48,8 @@ import net.sourceforge.pmd.util.OptionalBool;
 final class ClassStub implements JClassSymbol, AsmStub, AnnotationOwner {
 
     static final int UNKNOWN_ARITY = 0;
+    private static final int VISIBILITY_MASK = Opcodes.ACC_PUBLIC | Opcodes.ACC_PROTECTED | Opcodes.ACC_PRIVATE;
+    private static final int NO_INNER_CLASS_VISIBILITY = -1;
 
     private final AsmSymbolResolver resolver;
 
@@ -56,6 +58,7 @@ final class ClassStub implements JClassSymbol, AsmStub, AnnotationOwner {
     // all the following are lazy and depend on the parse lock
 
     private int accessFlags;
+    private int innerClassVisibility = NO_INNER_CLASS_VISIBILITY;
 
     private EnclosingInfo enclosingInfo;
     private LazyClassSignature signature;
@@ -197,23 +200,34 @@ final class ClassStub implements JClassSymbol, AsmStub, AnnotationOwner {
             Also ACC_SUPER conflicts with ACC_SYNCHRONIZED, which
             Modifier.toString would reflect.
 
-            Since the differences are disjoint we can just OR the two
-            sets of flags.
+            The non-visibility differences are disjoint, so they can be
+            combined with OR. Visibility needs to be merged separately:
+            bytecode processors may produce conflicting InnerClasses
+            entries, and class stubs can be populated in different orders.
          */
-        final int visibilityMask = Opcodes.ACC_PUBLIC | Opcodes.ACC_PROTECTED | Opcodes.ACC_PRIVATE;
-        int myAccess = this.accessFlags;
         if (fromClassInfo) {
             // we don't care about ACC_SUPER and it conflicts
             // with ACC_SYNCHRONIZED
             accessFlags = accessFlags & ~Opcodes.ACC_SUPER;
-        } else if ((myAccess & Opcodes.ACC_PUBLIC) != 0
-            && (accessFlags & visibilityMask) != Opcodes.ACC_PUBLIC) {
-            // ClassInfo mentions ACC_PUBLIC even if the real
-            // visibility is protected or private
-            // We remove the public to avoid a "public protected" or "public private" combination
-            myAccess = myAccess & ~Opcodes.ACC_PUBLIC;
+            if (innerClassVisibility != NO_INNER_CLASS_VISIBILITY) {
+                // InnerClasses contains the source-level visibility, even
+                // when it arrived before the ClassInfo for this class.
+                accessFlags = accessFlags & ~VISIBILITY_MASK;
+            }
+        } else {
+            int newVisibility = normalizeVisibility(accessFlags);
+            if (innerClassVisibility == NO_INNER_CLASS_VISIBILITY) {
+                innerClassVisibility = newVisibility;
+            } else {
+                innerClassVisibility = mostRestrictiveVisibility(innerClassVisibility, newVisibility);
+            }
+
+            // Replace any visibility read from ClassInfo or another
+            // InnerClasses entry, instead of accumulating incompatible flags.
+            this.accessFlags = this.accessFlags & ~VISIBILITY_MASK;
+            accessFlags = (accessFlags & ~VISIBILITY_MASK) | innerClassVisibility;
         }
-        this.accessFlags = myAccess | accessFlags;
+        this.accessFlags = this.accessFlags | accessFlags;
 
         // setModifiers is called multiple times: once from ClassFile structure (fromClassInfo==true)
         // and additionally from InnerClasses attribute (fromClassInfo==false)
@@ -227,6 +241,29 @@ final class ClassStub implements JClassSymbol, AsmStub, AnnotationOwner {
                 this.recordComponents = new ArrayList<>();
             }
         }
+    }
+
+    private static int normalizeVisibility(int accessFlags) {
+        int visibility = accessFlags & VISIBILITY_MASK;
+        if ((visibility & Opcodes.ACC_PRIVATE) != 0) {
+            return Opcodes.ACC_PRIVATE;
+        } else if (visibility == 0) {
+            return 0; // package-private
+        } else if ((visibility & Opcodes.ACC_PROTECTED) != 0) {
+            return Opcodes.ACC_PROTECTED;
+        }
+        return Opcodes.ACC_PUBLIC;
+    }
+
+    private static int mostRestrictiveVisibility(int first, int second) {
+        if (first == Opcodes.ACC_PRIVATE || second == Opcodes.ACC_PRIVATE) {
+            return Opcodes.ACC_PRIVATE;
+        } else if (first == 0 || second == 0) {
+            return 0; // package-private
+        } else if (first == Opcodes.ACC_PROTECTED || second == Opcodes.ACC_PROTECTED) {
+            return Opcodes.ACC_PROTECTED;
+        }
+        return Opcodes.ACC_PUBLIC;
     }
 
     void setEnclosingInfo(ClassStub outer, boolean localOrAnon, @Nullable String methodName, @Nullable String methodDescriptor) {

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ClassStub.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ClassStub.java
@@ -202,8 +202,10 @@ final class ClassStub implements JClassSymbol, AsmStub, AnnotationOwner {
 
             The non-visibility differences are disjoint, so they can be
             combined with OR. Visibility needs to be merged separately:
-            bytecode processors may produce conflicting InnerClasses
-            entries, and class stubs can be populated in different orders.
+            Bytecode processors (e.g. Android Proguard/R8) may produce
+            conflicting InnerClasses entries (widening access modifiers to
+            allow easier optimizations), and class stubs can be
+            populated in different orders.
          */
         if (fromClassInfo) {
             // we don't care about ACC_SUPER and it conflicts

--- a/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ClassStubTest.java
+++ b/pmd-java/src/test/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ClassStubTest.java
@@ -16,12 +16,15 @@ import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
+import java.io.ByteArrayInputStream;
 import java.lang.reflect.Modifier;
 import java.util.List;
 
 import org.checkerframework.checker.nullness.qual.NonNull;
 import org.jetbrains.annotations.NotNull;
 import org.junit.jupiter.api.Test;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.Opcodes;
 import org.pcollections.PSet;
 
 import net.sourceforge.pmd.lang.java.JavaParsingHelper;
@@ -37,6 +40,9 @@ import net.sourceforge.pmd.lang.java.types.TypeSystem;
 import net.sourceforge.pmd.util.CollectionUtil;
 
 class ClassStubTest {
+    private static final String R8_OUTER_INTERNAL_NAME = "testdata/R8Outer";
+    private static final String R8_INNER_INTERNAL_NAME = R8_OUTER_INTERNAL_NAME + "$Inner";
+
     // while parsing the annotation type, ClassStub's parseLock.ensureParsed()
     // is called multiple times, reentering the parselock while the status is
     // still BEING_PARSED.
@@ -140,6 +146,22 @@ class ClassStubTest {
         assertThat(innerOfObj, hasProperty("modifiers", equalTo(Modifier.PRIVATE | Modifier.STATIC)));
     }
 
+    @Test
+    void testLoadR8InnerClassWithConflictingVisibility() {
+        TypeSystem ts = TypeSystem.usingClassLoaderClasspath(JavaParsingHelper.class.getClassLoader());
+        AsmSymbolResolver resolver = (AsmSymbolResolver) ts.bootstrapResolver();
+        Loader loader = new Loader.StreamLoader(R8_INNER_INTERNAL_NAME,
+                                                new ByteArrayInputStream(r8InnerClass()));
+        ClassStub inner = new ClassStub(resolver, R8_INNER_INTERNAL_NAME, loader,
+                                        ClassStub.UNKNOWN_ARITY);
+
+        // Simulate the outer class being parsed first. Parsing the inner stub then adds
+        // public from both its ClassInfo and its own InnerClasses entry.
+        inner.setModifiers(Opcodes.ACC_PRIVATE, false);
+
+        assertEquals(Modifier.PRIVATE, inner.getModifiers());
+    }
+
 
     @Test
     void testLoadAnonClassFromEnum() {
@@ -195,6 +217,15 @@ class ClassStubTest {
 
     private static @NonNull JClassSymbol loadScalaClass(TypeSystem typeSystem, String simpleName) {
         return loadClassInPackage("net.sourceforge.pmd.lang.java.symbols.scalaclasses", simpleName, typeSystem);
+    }
+
+    private static byte[] r8InnerClass() {
+        ClassWriter writer = new ClassWriter(0);
+        writer.visit(Opcodes.V1_8, Opcodes.ACC_PUBLIC | Opcodes.ACC_SUPER,
+                     R8_INNER_INTERNAL_NAME, null, "java/lang/Object", null);
+        writer.visitInnerClass(R8_INNER_INTERNAL_NAME, R8_OUTER_INTERNAL_NAME, "Inner", Opcodes.ACC_PUBLIC);
+        writer.visitEnd();
+        return writer.toByteArray();
     }
 
     private static @NonNull JClassSymbol loadTestDataClass(TypeSystem typeSystem, String simpleName) {

